_WKContextMenuElementInfo should expose a _WKHitTestResult property

jeffmapple opened the pull request:
#### 194503237785a3e8e926acfb1dbadb179700cd56
<pre>
_WKContextMenuElementInfo should expose a _WKHitTestResult property
<a href="https://bugs.webkit.org/show_bug.cgi?id=250115">https://bugs.webkit.org/show_bug.cgi?id=250115</a>
&lt;rdar://103895210&gt;

Reviewed by Alex Christensen.

* Source/WebKit/Shared/API/APIObject.h:
Added ContextMenuElementInfoMac.

* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
Added support for _WKContextMenuElementInfo.

* Source/WebKit/UIProcess/API/APIContextMenuElementInfoMac.h:
Added. Unfortunately, there is already an iOS-only APIContextMenuElementInfo class and a
corresponding WKContextMenuElementInfo Objective-C class, so I had to use the Mac suffix here.

* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.h:
Declare a new hitTestResult property.

* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
This now wraps API::ContextMenuElementInfoMac.
(-[_WKContextMenuElementInfo dealloc]):
(-[_WKContextMenuElementInfo hitTestResult]):
Create a _WKHitTestResult from WebHitTestResultData.
(-[_WKContextMenuElementInfo _apiObject]):

* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfoInternal.h:
Added, wraps the API:APIContextMenuElementInfoMac class.

* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::ContextMenuClient::menuFromProposedMenu):
Use the hit test result when creating the _WKContextMenuElementInfo.
Clean up some style checker errors.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Added APIContextMenuElementInfoMac.h and _WKContextMenuElementInfoInternal.h.

* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST):
Added a test that just verifies the _WKContextMenuElementInfo hitTestResult property is set.

Canonical link: <a href="https://commits.webkit.org/258624@main">https://commits.webkit.org/258624@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e4d3015554ef6fa9a48daec522290ab8d9c23f97

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102529 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11661 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35563 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111796 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172013 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106497 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12668 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2563 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94803 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109506 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108308 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/37363 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24428 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5119 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25847 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5280 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2298 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11299 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45338 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5926 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7011 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
```diff
@@ -24,12 +24,22 @@
+    if (!(self = [super init]))
+        return nil;
+
+    _hitTestResult = hitTestResult;
```
I'm not sure this properly retains and releases the _WKHitTestResult to give it the right lifetime.  Implicit stuff like this in non-ARC ObjC makes me a little scared.

jeffmapple replied:
I think I've addressed this now that I'm using an API object.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfoInternal.h:
```diff
@@ -0,0 +1,43 @@
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class _WKHitTestResult;
```
I think this is redundant.

jeffmapple replied:
Can you elaborate? Don't I need to forward declare this class before defining the property below?

achristensen07 replied:
You do in _WKContextMenuElementInfo.h which this includes.

jeffmapple replied:
Maybe this is more of a philosophical issue, but shouldn't we try to make headers self-contained? I typically don't rely on non-obvious forward declarations that an included header might have. But I'll remove it if you feel strongly about it.

achristensen07 replied:
I don't feel strongly about it.

jeffmapple replied:
Actually, this becomes moot once I implement your other suggestion to use an API object.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfoInternal.h:
```diff
@@ -0,0 +1,43 @@
+WK_CLASS_AVAILABLE(macos(WK_MAC_TBA))
+@interface _WKContextMenuElementInfo ()
+
+- (instancetype)_initWithHitTestResult:(_WKHitTestResult *)hitTestResult;
```
Our preferred pattern for API objects is to have the internal header have a declaration that says the ObjC object contains a API::ObjectStorage of some type.

jeffmapple replied:
Done!

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -239,7 +239,8 @@
 
-    auto contextMenuElementInfo = adoptNS([[_WKContextMenuElementInfo alloc] init]);
+    auto apiHitTestResult = API::HitTestResult::create(data);
+    auto contextMenuElementInfo = adoptNS([[_WKContextMenuElementInfo alloc] _initWithHitTestResult:wrapper(apiHitTestResult.get())]);
```
I think this should look more like this:
auto contextMenuElementInfo = API::ContextMenuElementInfoMac::create(data);
then below we can use wrapper(contextMenuElementInfo)

jeffmapple replied:
Done!

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
```diff
@@ -35,6 +41,28 @@ - (id)copyWithZone:(NSZone *)zone
+- (_WKHitTestResult *)hitTestResult
+{
+    auto& hitTestResultData = _contextMenuElementInfoMac->hitTestResultData();
+    auto apiHitTestResult = API::HitTestResult::create(hitTestResultData);
```
This looks like nonatomic, readonly, copy to me instead of just nonatomic, readonly

jeffmapple replied:
Fixed.

jeffmapple committed:
_WKContextMenuElementInfo should expose a _WKHitTestResult property
https://bugs.webkit.org/show_bug.cgi?id=250115
<rdar://103895210>

Reviewed by Alex Christensen.

* Source/WebKit/Shared/API/APIObject.h:
Added ContextMenuElementInfoMac.

* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
Added support for _WKContextMenuElementInfo.

* Source/WebKit/UIProcess/API/APIContextMenuElementInfoMac.h:
Added. Unfortunately, there is already an iOS-only APIContextMenuElementInfo class and a
corresponding WKContextMenuElementInfo Objective-C class, so I had to use the Mac suffix here.

* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.h:
Declare a new hitTestResult property.

* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
This now wraps API::ContextMenuElementInfoMac.
(-[_WKContextMenuElementInfo dealloc]):
(-[_WKContextMenuElementInfo hitTestResult]):
Create a _WKHitTestResult from WebHitTestResultData.
(-[_WKContextMenuElementInfo _apiObject]):

* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfoInternal.h:
Added, wraps the API:APIContextMenuElementInfoMac class.

* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::ContextMenuClient::menuFromProposedMenu):
Use the hit test result when creating the _WKContextMenuElementInfo.
Clean up some style checker errors.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Added APIContextMenuElementInfoMac.h and _WKContextMenuElementInfoInternal.h.

* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST):
Added a test that just verifies the _WKContextMenuElementInfo hitTestResult property is set.

Canonical link: https://commits.webkit.org/258624@main
